File: backend/sdk/php/src/Services/Endpoints.php
<?php

declare(strict_types=1);

namespace WAAS\Services;

use WAAS\Client;
use WAAS\Models\{CreateEndpointRequest, WebhookEndpoint};

class Endpoints
{
    public function __construct(private readonly Client $client) {}

    public function create(CreateEndpointRequest $request): WebhookEndpoint
    {
        $response = $this->client->post('/webhooks/endpoints', $request->toArray());
        return WebhookEndpoint::fromArray($response);
    }

    public function get(string $endpointId): WebhookEndpoint
    {
        $response = $this->client->get("/webhooks/endpoints/{$endpointId}");
        return WebhookEndpoint::fromArray($response);
    }

    /**
     * @return array{endpoints: WebhookEndpoint[], total: int, page: int, per_page: int}
     */
    public function list(int $page = 1, int $perPage = 20): array
    {
        $response = $this->client->get("/webhooks/endpoints?page={$page}&per_page={$perPage}");
        return [
            'endpoints' => array_map(
                fn($e) => WebhookEndpoint::fromArray($e),
                $response['endpoints'] ?? []
            ),
            'total' => $response['total'] ?? 0,
            'page' => $response['page'] ?? $page,
            'per_page' => $response['per_page'] ?? $perPage,
        ];
    }

    /**
     * @param array<string, mixed> $attributes
     */
    public function update(string $endpointId, array $attributes): WebhookEndpoint
    {
        $response = $this->client->patch("/webhooks/endpoints/{$endpointId}", $attributes);
        return WebhookEndpoint::fromArray($response);
    }

    public function delete(string $endpointId): void
    {
        $this->client->delete("/webhooks/endpoints/{$endpointId}");
    }

    public function rotateSecret(string $endpointId): string
    {
        $response = $this->client->post("/webhooks/endpoints/{$endpointId}/rotate-secret");
        return $response['secret'];
    }
}

File: backend/sdk/ruby/lib/waas/services/endpoints.rb
# frozen_string_literal: true

module WAAS
  module Services
    class Endpoints
      def initialize(client)
        @client = client
      end

      def create(url:, custom_headers: nil, retry_config: nil)
        body = { url: url }
        body[:custom_headers] = custom_headers if custom_headers
        body[:retry_config] = retry_config if retry_config

        response = @client.post("/webhooks/endpoints", body)
        Models::WebhookEndpoint.from_hash(response)
      end

      def get(endpoint_id)
        response = @client.get("/webhooks/endpoints/#{endpoint_id}")
        Models::WebhookEndpoint.from_hash(response)
      end

      def list(page: 1, per_page: 20)
        response = @client.get("/webhooks/endpoints?page=#{page}&per_page=#{per_page}")
        {
          endpoints: (response["endpoints"] || []).map { |e| Models::WebhookEndpoint.from_hash(e) },
          total: response["total"],
          page: response["page"],
          per_page: response["per_page"]
        }
      end

      def update(endpoint_id, **attributes)
        response = @client.patch("/webhooks/endpoints/#{endpoint_id}", attributes)
        Models::WebhookEndpoint.from_hash(response)
      end

      def delete(endpoint_id)
        @client.delete("/webhooks/endpoints/#{endpoint_id}")
        true
      end

      def rotate_secret(endpoint_id)
        response = @client.post("/webhooks/endpoints/#{endpoint_id}/rotate-secret")
        response["secret"]
      end
    end
  end
end
